Add native polygonstolines algorithm

m-kuhn opened the pull request:
A 1:1 port of the existing Python polygons to lines algorithm.
Side product of a very confusing sip memory management debug session.

* [x] Unit Tests

m-kuhn committed:
Add native polygonstolines algorithm

m-kuhn committed:
Use iterator for parts

nirvn commented:
@m-kuhn , did you ever find what was wrong at the end?

nyalldawson commented on src/analysis/processing/qgsalgorithmpolygonstolines.cpp:
```diff
@@ -0,0 +1,151 @@
+  }
+  else if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( geom ) )
+  {
+    rings.append( polygon->exteriorRing()->clone() );
```
Do we need to guard against null exterior ring ? (i think we should).

m-kuhn replied:
Also added a test for that

nyalldawson commented:
I'm definitely in favour of python -> C++ ports!

m-kuhn commented:
> @m-kuhn , did you ever find what was wrong at the end?

See here: https://www.riverbankcomputing.com/pipermail/pyqt/2019-January/041251.html

The most probable suspicion is that a wrapper around a QgsAbstractGeometry (QgsPoint) pointer is cached in some Python code while the underlying object gets deleted and a new one (QgsMultiPolygon) created at the same address and the existing (QgsPoint) wrapper reused.

Conclusion: never cache a QgsAbstractGeometry pointer in Python, always cache the complete QgsGeometry (or clone() the QgsAbstractGeometry but I don't see the benefit of doing that).

m-kuhn committed:
Guard against nullptr

m-kuhn commented:
Not really, it's just a guideline python devs need to follow

m-kuhn committed:
Fix QgsGeometryCollection (and subclasses) `equal`

The previous behavior was to check if all parts of the collection are exactly the same.
The new check will check if all parts are equal (same vertice count, same vertice order, same vertice values).

m-kuhn committed:
Correctly assign output type from input type

m-kuhn committed:
Add tests for native polygonstolines algorithm

m-kuhn commented:
Added some boilerplate code for quick and dirty feature based algorithm tests
Includes a forward port of #8913

m-kuhn committed:
Add a nullptr guard in the geometry equality check

m-kuhn committed:
Minor code cleanup

nirvn commented:
I'd say +1 for 3.6; straightforward porting to c++ to fix a SIP issue is bug fixing. The extra benefits are just a cherry on top of the fix :)

nyalldawson commented:
+1 to 3.6

nyalldawson commented:
@m-kuhn 
I'm not sure Phil's interpretation of the issue is correct here. See https://issues.qgis.org/issues/20661#note-2 for another test case, which definitely **isn't** being cached.

m-kuhn commented:
>  I'm not sure Phil's interpretation of the issue is correct here. See https://issues.qgis.org/issues/20661#note-2 for another test case, which definitely isn't being cached.

Ouch, not sure what to make of that...
